contracts/programs/keystone-forwarder/src/lib.rs

```diff
@@ -0,0 +1,504 @@
+         .try_into()
+         .map_err(|_| ForwarderError::UnauthorizedSigner)?;
+
+         let index = oracles_config
+             .signer_addresses.binary_search_by(|addr| signer_eth_address.cmp(addr))
+             .map_err(|_| ForwarderError::UnauthorizedSigner)?;
+
+         uniques |= 1 << index;
+     };
+
+     require!(uniques.count_ones() as usize == total_signatures_len, ForwarderError::DuplicateSignatures);
+
+     Ok(())
+}
+
+
+fn set_oracles_config(oracles_config: &mut Account<OraclesConfig>, don_id: u32, config_version: u32, f: u8, signer_addresses: Vec<EthAddress>) -> Result<()> {
+    require!(signer_addresses.len() <= MAX_REPORT_SIGNERS.into(), ForwarderError::MaxSignersLimit);
+
+    let mut prev_signer: EthAddress = [0u8; 20];
+
+    for &curr_signer in signer_addresses.iter() {
+        // will also fail if there is a duplicate signer
+        require!(curr_signer > prev_signer, ForwarderError::SignersNotSortedInIncreasingOrder);
+
+        prev_signer = curr_signer;
+    }
+
+    oracles_config.config_id = ((don_id as u64) << 32) | (config_version as u64);
+    oracles_config.f = f;
+    oracles_config.signer_addresses = signer_addresses;
+
+    Ok(())
+}
+
+#[account]
+#[derive(Default)]
+pub struct ForwarderState {
+    version: u8,
+    authority_nonce: u8, // bump
+    owner: Pubkey,
+    proposed_owner: Pubkey,
+}
+
+
+impl ForwarderState {
+    // version + authority_nonce + owner + proposed_owner + vec prefix for config vec
+    pub const SPACE: usize = 1 + 1 + 4 + 4 + 32;
+}
+
+#[derive(Accounts)]
+pub struct Initialize<'info> {
+    // the account is not a PDA but it is initialized by the program
+    #[account(      
+        init,
+        payer = owner,
+        space = 8 + ForwarderState::SPACE
+    )]
+    pub state: Account<'info, ForwarderState>,
+    #[account(mut)]
+    pub owner: Signer<'info>,
+
+    pub system_program: Program<'info, System>,
+}
+
+#[derive(Accounts)]
+pub struct TransferOwnership<'info> {
+    #[account(mut)]
+    pub state: Account<'info, ForwarderState>,
+
+    #[account(address = state.owner @ AuthError::Unauthorized)]
+    pub current_owner: Signer<'info>,
+}
+
+#[derive(Accounts)]
+pub struct AcceptOwnership<'info> {
+    #[account(mut)]
+    pub state: Account<'info, ForwarderState>,
+
+    #[account(address = state.proposed_owner @ AuthError::Unauthorized)]
+    pub proposed_owner: Signer<'info>,
+}
+
+// combination of solidity's OracleSet and the configId mapping
+#[account]
+#[derive(Default)]
+pub struct OraclesConfig {
+    config_id: u64,
+    f: u8,
+    signer_addresses: Vec<EthAddress>, // maybe think about doing ECDSA over ed25519?
+} 
+
+impl OraclesConfig {
+    pub const INIT_SPACE: usize = 8 + 1 + 4;
+
+    pub fn space_with_signers(num_signers: usize) -> usize {
+        Self::INIT_SPACE + num_signers*20
+    }
+}
+
+fn get_config_id(don_id: u32, config_version: u32) -> u64 {
+    ((don_id as u64) << 32) | (config_version as u64)
```

Comment:
Config id could technically just be a struct with two u32 fields, it would have the same underlying size